ScheduledJob kubectl

soltysh opened the pull request:
@erictune the last part (kubectl) of the ScheduledJob, as usual builds on top of previous PRs, so only last 2 commits matter (the kubectl one and storage leftovers).

``` release-note
* Introducing ScheduledJobs as described in [the proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/scheduledjob.md) as part of `batch/v2alpha1` version (experimental feature).
```

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

soltysh committed:
ScheduledJob kubectl changes

soltysh committed:
ScheduledJob storage leftovers

soltysh commented:
@lavalamp I'm having some problems with initiating storage for scheduledjobs, iow. the `batch/v2alpha1` groups, see the following error I'm getting. What am I missing, got some ideas?
```
I0518 15:38:12.822156   25767 storage_factory.go:217] storing {batch scheduledjobs} in batch/v1, reading as batch/__internal from { {0xc82037b6c0 0xc82037b730} /registry [http://127.0.0.1:4001]    false 50000}
panic: storage codec doesn't seem to match given type: Internal type not encodable: no kind "ScheduledJob" is registered for version "batch/v1"

goroutine 1 [running]:
panic(0x1f0f1a0, 0xc820488a10)
        /usr/lib/golang/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/storage.NewCacherFromConfig(0x64, 0x7f524108c398, 0xc8201a3400, 0x7f524108c360, 0x41b9228, 0x2b03cc0, 0xc820183400, 0x2b5c730, 0xe, 0xc820570bc0, ...)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/cacher.go:156 +0x817
k8s.io/kubernetes/pkg/storage.NewCacher(0x7f524108c398, 0xc8201a3400, 0x64, 0x7f524108c360, 0x41b9228, 0x7f52410888e0, 0xc820183400, 0x2b5c730, 0xe, 0x7f52410c37b8, ...)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/cacher.go:142 +0x185
k8s.io/kubernetes/pkg/registry/generic/registry.StorageWithCacher(0x7f524108c398, 0xc8201a3400, 0x64, 0x7f52410888e0, 0xc820183400, 0x2b5c730, 0xe, 0x7f52410c37b8, 0xc820570ba0, 0x32911e8, ...)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/registry/storage_factory.go:36 +0x101
k8s.io/kubernetes/pkg/registry/scheduledjob/etcd.NewREST(0x7f524108c398, 0xc8201a3400, 0x3290fa8, 0x1, 0x3290fa8, 0x1)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/scheduledjob/etcd/etcd.go:41 +0x1cc
k8s.io/kubernetes/pkg/master.(*Master).getBatchResources(0xc8203ad930, 0xc820332fa0, 0x2b4dd78, 0x5, 0x2b92c90, 0x8, 0xc82037b260)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:844 +0x733
k8s.io/kubernetes/pkg/master.(*Master).InstallAPIs(0xc8203ad930, 0xc820332fa0)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:331 +0x2751
k8s.io/kubernetes/pkg/master.New(0xc820332fa0, 0x0, 0x0, 0x0)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:173 +0x189
k8s.io/kubernetes/cmd/kube-apiserver/app.Run(0xc8202c7680, 0x0, 0x0)
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:239 +0x2e3c
main.main()
        /home/maszulik/workspace/k8s/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/apiserver.go:50 +0x121
```

From my debugging it looks like the storagefactory is picking the wrong codec for ScheduledJob, do you have some thoughts where should I register one to be properly picked up?

lavalamp commented:
I have a long queue right now so sorry for the cryptic answer-- you need to somehow override the storage version for just the ScheduledJob resource to be v2alpha1 since it doesn't exist in v1. Not sure if the existing flag is going to make that easy or not.

soltysh commented:
@lavalamp thanks, will try to poke around and see what I can come up with.

soltysh committed:
Forced using batch/v2alpha1 for storing ScheduledJob

soltysh commented:
@lavalamp I've noticed there's a [SetResourceEncoding](https://github.com/kubernetes/kubernetes/blob/7e1089bb75328fef12c7a48a3e531f7a0d4b4c14/pkg/genericapiserver/resource_encoding_config.go#L70) which is not used currently but does exactly what you've mentioned. Iow. it allows specifying a storage version for specific resource, it's not exposed, though. Mind having a look and letting me know if it's more or less what you'd expect to look like?

soltysh commented:
@erictune @lavalamp with the current state I am able to get & create scheduledjobs using curl, but trying to use kubectl fails everytime with `error when creating "scheduledjob.yaml": no kind "ScheduledJob" is registered for version "batch/v1"` even when using `kubectl get scheduledjobs.v2alpha1.batch` which should reach the right endpoint. Any ideas?

lavalamp commented:
@soltysh What's the file you're creating look like? it needs to have batch/v2alpha1 as the apiVersion?

Also presumably you built an new kubectl?

soltysh commented:
@lavalamp this is the resource I'm trying to create: https://gist.github.com/soltysh/a917d4aac4cbccbff690117c0fc99a6c, but I still hit error: `error when creating "scheduledjob.yaml": no kind "ScheduledJob" is registered for version "batch/v1`

soltysh commented:
Re-applying lgtm label after fixing govet error.

soltysh commented:
The failure seems suspicious, as if not the latest code is in place. I've rebased it and repushed it. Will re-apply the lgtm label once bot removes it.

soltysh commented:
There's some integration flake, I'm working on. It has to do with storage for `batch/v2alpha1` which I thought is addressed properly, but somehow in integration locally I'm getting different results than in gke. Need to dig deeper into it.

goltermann commented:
@soltysh  This PR has the v1.3 milestone tag applied, but hasn't been touched in a week or more. We will cut Beta and branch on June 10. Please either close, finish off, get an LGTM or rebase this as needed to make sure it makes its way into 1.3.

soltysh committed:
ScheduledJob suspend being pointer leftovers

soltysh committed:
ScheduledJob suspend being pointer leftovers - generated changes

soltysh commented:
@erictune I've updated the PR and found additional missing pieces. There's still a problem when trying to create a scheduledjob (even with Clayton's PR). I'm trying to debug the problem atm.

soltysh commented:
So it looks like kubectl is always using [batchClient](https://github.com/kubernetes/kubernetes/blob/339c6924b2d8a8d2a5713d3c9964d7ebe0dcdec1/pkg/client/unversioned/helper.go#L77) instead of the [discovery one](https://github.com/kubernetes/kubernetes/blob/339c6924b2d8a8d2a5713d3c9964d7ebe0dcdec1/pkg/client/unversioned/helper.go#L60). If I force using `batch/v2alpha1` client specifically it'll work.

soltysh commented:
After further investigation I'm suspecting this boils down to [serializer](https://github.com/kubernetes/kubernetes/blob/339c6924b2d8a8d2a5713d3c9964d7ebe0dcdec1/pkg/client/restclient/client.go#L144) which is using only single GroupVersion encoder and decoder.

soltysh commented:
In the end it comes to #25676, which is required for kubectl to work properly with ScheduledJobs. I'm currently testing if with those changes this will work..

smarterclayton commented:
I really doubt that is the only way to fix your issue.  The
RestMapping used to call create should be able to return v2alpha1.
Please include the RestMapping returned for create here.

soltysh commented:
@smarterclayton the problem here is not with the RestMapping, which iirc does know how to handle `batch/v2alph1` but with the [serializer](https://github.com/kubernetes/kubernetes/blob/339c6924b2d8a8d2a5713d3c9964d7ebe0dcdec1/pkg/client/restclient/client.go#L144) which is created for [specific version](https://github.com/kubernetes/kubernetes/blob/339c6924b2d8a8d2a5713d3c9964d7ebe0dcdec1/pkg/client/unversioned/helper.go#L77). After discussing this topic with @deads2k I have two options, go with @krousey dynamic PR or fix the serializer to handle multiple versions. Unless I'm missing 3rd option.
What wonders me though, is why `batch/v2alpha1`->`batch/__internal` conversion works ok (when creating) but back does not, or rather is done differently. I'll try to get you on IRC to get some discussion around this topic.

soltysh commented:
I've spent a few minutes to cherry-pick @krousey PR and updating it and creating SJ works, unfortunately getting/listing them no :(

deads2k commented:
`ClientForMapping` chooses a `RESTClient` based on the group, not the groupversion.  because the config backing that `RESTClient` is pinned to a particular group/version, you'd have to create a second kind of client to accommodate it or have an option on the client to change its groupversion association.

soltysh commented:
Yeah, that's an option I'd prefer to avoid, because it's not salable. And soon I'm hoping there will be more groups supporting multiversions.

soltysh commented:
OK, there's the 3rd option, I've spoke with Clayton a minute ago. Will create a separate PR to fix it.

deads2k commented:
> OK, there's the 3rd option, I've spoke with Clayton a minute ago. Will create a separate PR to fix it.

Was it too small to fit in the margin?

smarterclayton commented:
The logic in ClientForMapping in the factory is wrong.  We should be asking for ClientConfig and then parameterizing based on the incoming GV.

smarterclayton commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -343,46 +343,28 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
-			}
 			switch gvk.Group {
 			case api.GroupName:
+				mappingVersion := mapping.GroupVersionKind.GroupVersion()
```
All that is different between this and default should be that APIPath is "/api" vs "/apis"

soltysh commented:
This currently relies on https://github.com/kubernetes/kubernetes/pull/29187 being in first.

soltysh commented on pkg/kubectl/describe.go:
```diff
@@ -1194,6 +1203,92 @@ func describeJob(job *batch.Job, events *api.EventList) (string, error) {
+			fmt.Fprintf(out, "Last Schedule Time:\t<unset>\n")
+		}
+		printActiveJobs(out, "Active Jobs", scheduledJob.Status.Active)
+		if events != nil {
```
Reminder to myself, use `describerSettings.ShowEvents` to conditionally disable showing them. I think Jobs don't do it either.

soltysh replied:
It's done properly.

erictune commented:
@janetkuo Please review this.

janetkuo commented on pkg/genericapiserver/default_storage_factory_builder.go:
```diff
@@ -32,18 +32,30 @@ import (
 
-// Merges the given defaultAPIResourceConfig with the given storageEncodingOverrides.
-func mergeResourceEncodingConfigs(defaultResourceEncoding *DefaultResourceEncodingConfig, storageEncodingOverrides map[string]unversioned.GroupVersion) *DefaultResourceEncodingConfig {
+// Merges the given defaultResourceConfig with the given resourceEncodingOverrides.
```
The difference between `mergeResourceEncodingConfigs` and `mergeGroupEncodingConfigs` are hard to distinguish without looking at the code. Update the comment would help.

soltysh replied:
Updated.

janetkuo commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -139,6 +139,8 @@ func Run(s *options.APIServer) error {
 	storageFactory, err := genericapiserver.BuildDefaultStorageFactory(
 		s.StorageConfig, s.DefaultStorageMediaType, api.Codecs,
 		genericapiserver.NewDefaultResourceEncodingConfig(), storageGroupsToEncodingVersion,
+		// FIXME: that should be configurable
```
What should be configurable? Please clarify

janetkuo replied:
Is there an open issue/PR for this? If so put the number here too.

soltysh replied:
It's about making the `GroupVersionResource` overrides being configurable, I'll open a separate issue to address that.

janetkuo commented on pkg/api/testing/fuzzer.go:
```diff
@@ -169,6 +169,11 @@ func FuzzerFor(t *testing.T, version unversioned.GroupVersion, src rand.Source)
+		func(sj *batch.ScheduledJobSpec, c fuzz.Continue) {
+			c.FuzzNoCustom(sj)
+			suspend := c.RandBool()
+			sj.Suspend = &suspend
```
Do we also need to fuzz Schedule and StartingDeadlineSeconds?

soltysh replied:
Done.

janetkuo commented on pkg/apis/batch/v2alpha1/defaults.go:
```diff
@@ -46,4 +46,8 @@ func SetDefaults_ScheduledJob(obj *ScheduledJob) {
 	}
+	if obj.Spec.Suspend == nil {
+		obj.Spec.Suspend = new(bool)
+		*obj.Spec.Suspend = false
```
We don't actually need this line since it's defaulted to false in L50 (or did you just want to be more explicit here?)

soltysh replied:
Yeah, being explicit :wink:, will remove.

janetkuo commented on pkg/apis/batch/types.go:
```diff
@@ -209,7 +209,7 @@ type ScheduledJobSpec struct {
 	// Suspend flag tells the controller to suspend subsequent executions, it does
 	// not apply to already started executions.  Defaults to false.
-	Suspend *bool `json:"suspend"`
+	Suspend *bool `json:"suspend,omitempty"`
```
Just curious why use `*bool` here since it's an optional field and there's no different between this value be unset or false

soltysh replied:
If this is not a pointer user will be forced to define it, so it's not that much optional.

janetkuo replied:
Ah, I see!

janetkuo commented on pkg/client/unversioned/apps.go:
```diff
@@ -65,12 +65,10 @@ func setAppsDefaults(config *restclient.Config) error {
-	config.GroupVersion = &copyGroupVersion
-	//}
-
+	if config.GroupVersion == nil || !isEnabledVersion(apps.GroupName, config.GroupVersion) {
```
We do L68-71 in a lot of places. Shall we make it a function?

soltysh replied:
It's handled in #29187, already.

janetkuo commented on pkg/kubectl/cmd/run.go:
```diff
@@ -118,6 +121,7 @@ func addRunFlags(cmd *cobra.Command) {
 	cmd.Flags().String("service-generator", "service/v2", "The name of the generator to use for creating a service.  Only used if --expose is true")
 	cmd.Flags().String("service-overrides", "", "An inline JSON override for the generated service object. If this is non-empty, it is used to override the generated object. Requires that the object supply a valid apiVersion field.  Only used if --expose is true.")
 	cmd.Flags().Bool("quiet", false, "If true, suppress prompt messages.")
+	cmd.Flags().String("schedule", "", "A schedule the job should be run with.")
```
Mention that it should be in Cron format.

soltysh replied:
Done.

janetkuo commented on pkg/kubectl/describe.go:
```diff
@@ -1194,6 +1203,92 @@ func describeJob(job *batch.Job, events *api.EventList) (string, error) {
+
+	for i, job := range jobs {
+		if i != 0 {
+			fmt.Fprint(out, ",")
```
add a space between `,`

soltysh replied:
Done.

janetkuo commented on pkg/kubectl/resource_printer.go:
```diff
@@ -969,6 +972,50 @@ func printJobList(list *batch.JobList, w io.Writer, options PrintOptions) error
+		}
+	}
+
+	if scheduledJob.Status.LastScheduleTime != nil {
```
Change it to something like 
```go
lastSchedule := "<none>"
if scheduledJob.Status.LastScheduleTime != nil {
  lastSchedule = scheduledJob.Status.LastScheduleTime.Time.Format(time.RFC1123Z)
}
// Then Fprint the output

soltysh replied:
Done.

janetkuo commented on pkg/kubectl/resource_printer.go:
```diff
@@ -413,6 +413,7 @@ var podTemplateColumns = []string{"TEMPLATE", "CONTAINER(S)", "IMAGE(S)", "PODLA
 var replicationControllerColumns = []string{"NAME", "DESIRED", "CURRENT", "AGE"}
 var replicaSetColumns = []string{"NAME", "DESIRED", "CURRENT", "AGE"}
 var jobColumns = []string{"NAME", "DESIRED", "SUCCESSFUL", "AGE"}
+var scheduledJobColumns = []string{"NAME", "SCHEDULE", "SUSPEND", "ACTIVE", "LASTSCHEDULE"}
```
Change `LASTSCHEDULE` to `LAST-SCHEDULE`. By convention we use `-` as the word separator, see https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#output-conventions

soltysh replied:
Done.

janetkuo commented:
I had a few comments.

soltysh commented:
@janetkuo thanks! I'll update and address the comments once #29187 gets in.

janetkuo commented:
just a heads up that #29187 is in now

soltysh commented:
> just a heads up that #29187 is in now 

Yup, seen it. I'm currently updating this one. Should be ready in a few minutes.

soltysh commented:
Rebased and addressed all your comments @janetkuo . The only problem I'm struggling with is `hack/update-generated-protobuf.sh` is failing for me locally, but hopefully that won't be a problem with this PR. ptal

janetkuo commented:
Code looks good, but the `panic: storage codec doesn't seem to match given type: Internal type not encodable: no kind "ScheduledJob" is registered for version "batch/v1"` error is back again

soltysh commented:
> Code looks good, but the panic: storage codec doesn't seem to match given type: Internal type not encodable: no kind "ScheduledJob" is registered for version "batch/v1" error is back again

Of course it does... :sob: I'll try to figure it out later today.

soltysh commented:
This is a problem with integration tests that I need to fix to address multiversion tests.

soltysh commented:
@janetkuo the problem was that I've added `batchapi.v2alpha1.SchemeGroupVersion` in [DefaultAPIResourceConfigSource](https://github.com/kubernetes/kubernetes/blob/2ae16a2e6073a81a78dc1704894a8cf70b0bae42/pkg/master/master.go#L751) but I should not. I've removed it, but I've extended our integration tests with `batch/v2alpha1` version.

soltysh commented:
There's still one problem, already filled in issue #29998, but I think we'll have to deal with it as a followup. I'd like to move forward with the existing SJ PRs.

soltysh commented:
It looks like the rbac is not working as expected against multiversioned batch, I've created https://github.com/kubernetes/kubernetes/issues/30079 to track this one and I'm removing the last commit (adding `batch/v2alpha1` to integration) from this PR.

soltysh commented:
@erictune @janetkuo ptal
